flexmeasures/data/models/planning/storage.py

```diff
@@ -368,50 +386,345 @@ def build_device_soc_targets(
+    # B. Validation in the same time frame #
+    ########################################
+
+    # 1) min <= max
+    constraint_violations += validate_constraint(_constraints, "min(t) <= max(t)")
+
+    # 2) min <= equals
+    constraint_violations += validate_constraint(_constraints, "min(t) <= equals(t)")
+
+    # 3) equals <= max
+    constraint_violations += validate_constraint(_constraints, "equals(t) <= max(t)")
+
+    ##########################################
+    # C. Validation in different time frames #
+    ##########################################
+
+    _constraints["factor_w_wh(t)"] = resolution / timedelta(hours=1)
+    _constraints["min(t-1)"] = prepend_serie(_constraints["min(t)"], soc_min)
+    _constraints["equals(t-1)"] = prepend_serie(_constraints["equals(t)"], soc_at_start)
+    _constraints["max(t-1)"] = prepend_serie(_constraints["max(t)"], soc_max)
+
+    # 1) equals(t) - equals(t-1) <= derivative_max(t)
+    constraint_violations += validate_constraint(
+        _constraints, "equals(t) - equals(t-1) <= derivative_max(t) * factor_w_wh(t)"
+    )
+
+    # 2) derivative_min(t) <= equals(t) - equals(t-1)
+    constraint_violations += validate_constraint(
+        _constraints, "derivative_min(t) * factor_w_wh(t) <= equals(t) - equals(t-1)"
+    )
+
+    # 3) min(t) - max(t-1) <= derivative_max(t)
+    constraint_violations += validate_constraint(
+        _constraints, "min(t) - max(t-1) <= derivative_max(t) * factor_w_wh(t)"
+    )
+
+    # 4) max(t) - min(t-1) >= derivative_min(t)
+    constraint_violations += validate_constraint(
+        _constraints, "derivative_min(t) * factor_w_wh(t) <= max(t) - min(t-1)"
+    )
+
+    # 5) equals(t) - max(t-1) <= derivative_max(t)
+    constraint_violations += validate_constraint(
+        _constraints, "equals(t) - max(t-1) <= derivative_max(t) * factor_w_wh(t)"
+    )
+
+    # 6) derivative_min(t) <= equals(t) - min(t-1)
+    constraint_violations += validate_constraint(
+        _constraints, "derivative_min(t) * factor_w_wh(t) <= equals(t) - min(t-1)"
+    )
+
+    return constraint_violations
+
+
+def get_pattern_match_word(word: str) -> str:
+    """Get a regex pattern to match a word
+
+    The conditions to delimit a word are:
+      - start of line
+      - whitespace
+      - end of line
+      - word boundary
+      - arithmetic operations
+
+    :return: regex expression
+    """
+
+    regex = r"(^|\s|$|\b|\+|\-|\*|/\|\\)"
+
+    return regex + re.escape(word) + regex
+
+
+def validate_constraint(
+    constraints_df: pd.DataFrame, constraint_expression: str
+) -> list[dict]:
+    """Validate the feasibility of a given set of constraints.
+
+    :param constraints_df: DataFrame with the constraints
+    :param constraint_expression: inequality expression following pd.eval format.
+                                  No need to use the syntax `column` to reference
+                                  column, just use the column name.
```

Comment:
Question: how are we dealing with the `(t-1)` in the evaluated constraint expressions? Did you basically also already make it possible to evaluate things like comparing `(t)` with `(t-2)`?

Reply:
Good question! 

The first part of the function wraps the column names in the expression with accents `'column'` so that pandas eval can interpret them properly (e.g `max` doesn't work as is because it's interpret as the max **function**). For example, it wraps `max(t-1) + max(t)` in accents `'max(t-1)' + 'max(t)'`.

That way, if the column is available, we can use it in the `constraint_expression`.

Note: I don't know how to put them in GH, using `'`, instead.

Reply:
Yes, that's right. 

The alternative would be to have one expression to show to the user and another one to compute the condition.

Reply:
Okay, thanks. I get it now. Then I suggest we create a follow-up issue to add extra validation rules, that expand case C from covering 1 step towards covering 1, 2, 3, ..., x, ..., n steps, where n is the number of steps in the scheduling horizon. For example, C.1) for `x=2` would become something like:

```
"equals(t) - equals(t-2) <= (derivative_max(t) + derivative_max(t-1)) * factor_w_wh(t)"
```

And generally, for `x` (but this probably still needs some love to get correctly parsed):

```
"equals(t) - equals(t-x) <= sum(derivative_max(t-k) for k in range(x)) * factor_w_wh(t)"
```

(Although `factor_w_wh(t)` can be considered a constant in our platform (i.e. not a function of `t`), I get why you just added it to the frame as a column.)

Reply:
Yeah, good idea! 

Maybe the cleanest would be to create a differentiation operator that can act on variables, we can discuss this further in the follow-up issue.

I think for the type of constraints that we have now, we still don't require the following constraint: 
```
"equals(t) - equals(t-2) <= (derivative_max(t) + derivative_max(t-1)) * factor_w_wh(t)"
```

The justification is the following, let me know if I'm getting something wrong:

let `e(t)` be the short name for `equals(t)` and `m(t)`  for `derivative_max(t)`:

```math
e(t) - e(t-2) = [e(t) - e(t-1)] + [e(t-1) - e(t-2)] \leq m(t) + m(t-1)
```
this follows directly from. 

```math
e(t) - e(t-1) \leq m(t)
```

Reply:
Ah, but what if there *is* no $e(t-1)$?

Reply:
I think that when there's no $e(t-1)$ this constraint can't be checked.

Reply:
Discussion continues in #704.